File: erpnext/manufacturing/page/bom_status/bom_status.js
frappe.pages['bom-status'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'وضعیت BOM - مصرف مواد',
        single_column: true
    });

    // اضافه کردن CSS برای استایل بهتر
    $(`
    <style>
        .bom-status-container {
            padding: 15px;
        }
        .bom-search-section {
            background: #f8f9fa;
            padding: 20px;
            border-radius: 8px;
            margin-bottom: 20px;
        }
        .bom-table {
            background: white;
            border-radius: 8px;
            overflow: hidden;
            box-shadow: 0 2px 4px rgba(0,0,0,0.1);
        }
        .bom-table table {
            margin-bottom: 0;
        }
        .bom-header-row {
            background: #007bff;
            color: white;
            font-weight: bold;
        }
        .bom-level-1 { background: #f8f9fa; }
        .bom-level-2 { background: #fff3cd; }
        .bom-level-3 { background: #d4edda; }
        .bom-level-4 { background: #cce7ff; }
        .bom-level-5 { background: #f0e6ff; }
        
        .expandable-row {
            cursor: pointer;
        }
        .expandable-row:hover {
            background: #e9ecef !important;
        }
        .expand-icon {
            margin-right: 5px;
            transition: transform 0.3s;
        }
        .expand-icon.expanded {
            transform: rotate(90deg);
        }
        .item-indent {
            padding-right: 20px;
        }
        .collapsed-row {
            display: none;
        }
        .search-controls {
            display: flex;
            gap: 15px;
            flex-wrap: wrap;
            align-items: center;
        }
        .control-group {
            flex: 1;
            min-width: 200px;
        }
        .btn-group {
            gap: 10px;
        }
    </style>
    `).appendTo('head');

    let container = $('<div class="bom-status-container">').appendTo(page.body);
    let currentData = [];
    let expandedRows = new Set();

    // ساخت بخش جستجو و کنترل‌ها
    let searchSection = $(`
        <div class="bom-search-section">
            <div class="search-controls">
                <div class="control-group">
                    <label>انتخاب BOM:</label>
                    <select class="form-control" id="bom-selector">
                        <option value="">همه BOM ها</option>
                    </select>
                </div>
                <div class="control-group">
                    <label>جستجو:</label>
                    <input type="text" class="form-control" id="search-input" placeholder="جستجو در BOM ها و آیتم‌ها...">
                </div>
                <div class="btn-group">
                    <button class="btn btn-primary" id="search-btn">جستجو</button>
                    <button class="btn btn-secondary" id="clear-btn">پاک کردن</button>
                    <button class="btn btn-success" id="expand-all-btn">باز کردن همه</button>
                    <button class="btn btn-warning" id="collapse-all-btn">بستن همه</button>
                </div>
            </div>
        </div>
    `).appendTo(container);

    let tableContainer = $('<div class="bom-table">').appendTo(container);

    // بارگذاری لیست BOM ها
    loadBOMList();

    // Event handlers
    $('#bom-selector').change(function() {
        let selectedBOM = $(this).val();
        if (selectedBOM) {
            loadBOMData('', selectedBOM);
        } else {
            loadBOMData();
        }
    });

    $('#search-btn').click(function() {
        let searchText = $('#search-input').val();
        let selectedBOM = $('#bom-selector').val();
        loadBOMData(searchText, selectedBOM);
    });

    $('#search-input').keypress(function(e) {
        if (e.which === 13) { // Enter key
            $('#search-btn').click();
        }
    });

    $('#clear-btn').click(function() {
        $('#search-input').val('');
        $('#bom-selector').val('');
        expandedRows.clear();
        loadBOMData();
    });

    $('#expand-all-btn').click(function() {
        expandAllRows();
    });

    $('#collapse-all-btn').click(function() {
        collapseAllRows();
    });

    function loadBOMList() {
        frappe.call({
            method: 'erpnext.manufacturing.page.bom_status.bom_status.get_all_boms_simple',
            callback: function(r) {
                if (r.message) {
                    let selector = $('#bom-selector');
                    selector.empty().append('<option value="">همه BOM ها</option>');
                    
                    r.message.forEach(bom => {
                        selector.append(`<option value="${bom.name}">${bom.name} - ${bom.item_name || bom.item_code}</option>`);
                    });
                }
            }
        });
    }

    function loadBOMData(searchText = '', selectedBOM = '') {
        frappe.call({
            method: 'erpnext.manufacturing.page.bom_status.bom_status.get_bom_status_simple',
            args: {
                search_text: searchText,
                selected_bom: selectedBOM
            },
            callback: function(r) {
                if (r.message) {
                    currentData = r.message;
                    renderBOMTable(r.message);
                }
            }
        });
    }

    function renderBOMTable(data) {
        if (!data || data.length === 0) {
            tableContainer.html('<div class="alert alert-info">هیچ داده‌ای یافت نشد</div>');
            return;
        }

        // اگر فقط لیست BOM ها برگردونده شده
        if (data[0] && data[0].type === 'bom_list') {
            renderBOMList(data[0].bom_list);
            return;
        }

        let html = `
        <table class="table table-bordered table-hover">
            <thead class="bom-header-row">
                <tr>
                    <th style="width: 30px;"></th>
                    <th>کد محصول</th>
                    <th>نام محصول</th>
                    <th>مقدار مصرفی</th>
                    <th>واحد</th>
                    <th>نرخ واحد</th>
                    <th>کل مبلغ</th>
                    <th>عملیات</th>
                </tr>
            </thead>
            <tbody>
        `;

        let currentBOM = '';
        
        data.forEach((item, index) => {
            if (item.type === 'bom_header') {
                html += renderBOMHeader(item);
                currentBOM = item.bom_name;
            } else if (item.type === 'bom_item') {
                html += renderBOMItem(item, index, currentBOM);
            }
        });

        html += '</tbody></table>';
        tableContainer.html(html);

        // اضافه کردن event handlers برای expand/collapse
        attachTableEvents();
    }

    function renderBOMList(bomList) {
        if (!bomList || bomList.length === 0) {
            tableContainer.html('<div class="alert alert-info">هیچ BOM فعالی یافت نشد</div>');
            return;
        }

        let html = `
        <table class="table table-bordered table-hover">
            <thead class="bom-header-row">
                <tr>
                    <th>کد BOM</th>
                    <th>کد محصول</th>
                    <th>نام محصول</th>
                    <th>عملیات</th>
                </tr>
            </thead>
            <tbody>
        `;

        bomList.forEach(bom => {
            html += `
            <tr>
                <td><strong>${bom.name}</strong></td>
                <td>${bom.item_code}</td>
                <td>${bom.item_name || bom.item_code}</td>
                <td>
                    <button class="btn btn-sm btn-primary view-bom-btn" data-bom="${bom.name}">
                        مشاهده جزئیات
                    </button>
                </td>
            </tr>
            `;
        });

        html += '</tbody></table>';
        tableContainer.html(html);

        // اضافه کردن event handler برای دکمه‌های مشاهده
        $('.view-bom-btn').click(function() {
            let bomName = $(this).data('bom');
            $('#bom-selector').val(bomName);
            loadBOMData('', bomName);
        });
    }

    function renderBOMHeader(item) {
        return `
        <tr class="bom-header-row">
            <td colspan="8">
                <strong>BOM: ${item.bom_name}</strong> | 
                محصول: ${item.bom_item} - ${item.bom_item_name || ''} | 
                کل هزینه: ${formatNumber(item.total_cost)} | 
                مقدار: ${item.quantity} ${item.uom}
            </td>
        </tr>
        `;
    }

    function renderBOMItem(item, index, currentBOM) {
        let indent = '&nbsp;&nbsp;'.repeat((item.level - 1) * 2);
        let levelClass = `bom-level-${Math.min(item.level, 5)}`;
        let rowId = `row-${currentBOM}-${index}`;
        let hasSubBOM = item.has_sub_bom == 1;
        let isExpanded = expandedRows.has(rowId);
        
        let expandIcon = '';
        let rowClass = levelClass;
        
        if (hasSubBOM) {
            expandIcon = `<i class="fa fa-chevron-right expand-icon ${isExpanded ? 'expanded' : ''}" data-row="${rowId}"></i>`;
            rowClass += ' expandable-row';
        }

        return `
        <tr class="${rowClass}" id="${rowId}" data-level="${item.level}" data-parent="${item.bom_name}">
            <td>${expandIcon}</td>
            <td class="item-indent">${indent}${item.item_code}</td>
            <td>${item.item_name || item.item_code}</td>
            <td class="text-center">${formatNumber(item.total_required_qty)}</td>
            <td class="text-center">${item.uom}</td>
            <td class="text-left">${formatNumber(item.rate)}</td>
            <td class="text-left">${formatNumber(item.total_amount)}</td>
            <td class="text-center">
                ${hasSubBOM ? `<button class="btn btn-xs btn-info expand-sub-bom" data-item="${item.item_code}">نمایش زیرمجموعه</button>` : '-'}
            </td>
        </tr>
        `;
    }

    function attachTableEvents() {
        // Event handler برای expand/collapse
        $('.expand-icon').click(function(e) {
            e.stopPropagation();
            let rowId = $(this).data('row');
            toggleRowExpansion(rowId);
        });

        $('.expandable-row').click(function() {
            let rowId = $(this).attr('id');
            toggleRowExpansion(rowId);
        });

        // Event handler برای نمایش زیرمجموعه
        $('.expand-sub-bom').click(function(e) {
            e.stopPropagation();
            let itemCode = $(this).data('item');
            // اینجا می‌تونید logic برای باز کردن BOM زیرمجموعه اضافه کنید
            frappe.msgprint(`نمایش زیرمجموعه برای ${itemCode} - این قابلیت در حال توسعه است`);
        });
    }

    function toggleRowExpansion(rowId) {
        if (expandedRows.has(rowId)) {
            expandedRows.delete(rowId);
            $(`#${rowId} .expand-icon`).removeClass('expanded');
        } else {
            expandedRows.add(rowId);
            $(`#${rowId} .expand-icon`).addClass('expanded');
        }
        
        // اینجا logic برای نمایش/مخفی کردن ردیف‌های فرزند
        // در صورت نیاز می‌تونید پیاده‌سازی کنید
    }

    function expandAllRows() {
        $('.expand-icon').each(function() {
            let rowId = $(this).data('row');
            expandedRows.add(rowId);
            $(this).addClass('expanded');
        });
    }

    function collapseAllRows() {
        expandedRows.clear();
        $('.expand-icon').removeClass('expanded');
    }

    function formatNumber(num) {
        if (!num || isNaN(num)) return '0';
        return parseFloat(num).toLocaleString('fa-IR', {
            minimumFractionDigits: 2,
            maximumFractionDigits: 2
        });
    }

    // بارگذاری اولیه داده‌ها
    loadBOMData();
}

File: erpnext/manufacturing/page/bom_status/bom_status.py
import frappe
from frappe.utils import flt

@frappe.whitelist()
def get_bom_status_simple(search_text="", selected_bom=""):
    """
    Simple BOM status with material consumption (max 3 levels)
    """
    try:
        if selected_bom:
            return get_bom_details_simple(selected_bom)
        elif search_text:
            return search_bom_items(search_text)
        else:
            return get_all_boms_simple()
    except Exception as e:
        frappe.log_error(f"Error in get_bom_status_simple: {str(e)}")
        return []

@frappe.whitelist() 
def get_all_boms_simple():
    """
    Get all active BOMs
    """
    query = """
    SELECT 
        name,
        item_name,
        total_cost,
        uom
    FROM `tabBOM`
    WHERE docstatus = 1 
      AND is_active = 1
    ORDER BY name
    LIMIT 100
    """
    
    boms = frappe.db.sql(query, as_dict=True)
    return [{"type": "bom_list", "bom_list": boms}]

def get_bom_details_simple(bom_name):
    """
    Get BOM details with 3 levels maximum
    """
    try:
        # گرفتن اطلاعات BOM اصلی
        bom_info = frappe.get_doc("BOM", bom_name)
        
        result = [{
            "type": "bom_header",
            "bom_name": bom_name,
            "bom_item": bom_info.item_code,
            "bom_item_name": bom_info.item_name,
            "total_cost": bom_info.total_cost or 0,
            "uom": bom_info.uom,
            "quantity": bom_info.quantity or 1
        }]
        
        # سطح 1: آیتم‌های مستقیم
        level1_items = get_bom_items(bom_name, 1)
        result.extend(level1_items)
        
        # سطح 2 و 3: زیرمجموعه‌ها
        for item in level1_items:
            if has_bom(item['item_code']):
                level2_items = get_bom_items_for_item(item['item_code'], 2, item['total_required_qty'])
                result.extend(level2_items)
                
                # سطح 3
                for item2 in level2_items:
                    if has_bom(item2['item_code']) and item2['level'] == 2:
                        level3_items = get_bom_items_for_item(item2['item_code'], 3, item2['total_required_qty'])
                        result.extend(level3_items)
        
        return result
        
    except Exception as e:
        frappe.log_error(f"Error in get_bom_details_simple: {str(e)}")
        return []

def get_bom_items(bom_name, level, multiplier=1):
    """
    Get items for a specific BOM
    """
    query = """
    SELECT 
        item.item_code,
        item.item_name,
        item.qty,
        item.uom,
        COALESCE(item.rate, 0) as rate,
        COALESCE(item.amount, 0) as amount,
        COALESCE(item_master.item_group, '') as item_group
    FROM `tabBOM Item` item
    LEFT JOIN `tabItem` item_master ON item_master.name = item.item_code
    WHERE item.parent = %s
    ORDER BY item.idx
    """
    
    items = frappe.db.sql(query, (bom_name,), as_dict=True)
    
    result = []
    for item in items:
        # فقط آیتم‌های گروه محصولات یا بدون گروه
        if not item['item_group'] or item['item_group'] == 'محصولات' or item['item_group'] == '':
            total_qty = item['qty'] * multiplier
            result.append({
                "type": "bom_item",
                "bom_name": bom_name,
                "item_code": item['item_code'],
                "item_name": item['item_name'] or item['item_code'],
                "qty": item['qty'],
                "total_required_qty": total_qty,
                "uom": item['uom'],
                "rate": item['rate'],
                "amount": item['amount'],
                "total_amount": total_qty * item['rate'],
                "level": level,
                "has_sub_bom": 1 if has_bom(item['item_code']) else 0,
                "item_group": item['item_group']
            })
    
    return result

def get_bom_items_for_item(item_code, level, multiplier):
    """
    Get BOM items for a specific item code
    """
    # پیدا کردن BOM برای این آیتم
    bom_query = """
    SELECT name FROM `tabBOM`
    WHERE item_code = %s 
      AND docstatus = 1 
      AND is_active = 1
    LIMIT 1
    """
    
    bom_result = frappe.db.sql(bom_query, (item_code,), as_dict=True)
    
    if bom_result:
        return get_bom_items(bom_result[0]['name'], level, multiplier)
    
    return []

def has_bom(item_code):
    """
    Check if item has an active BOM
    """
    query = """
    SELECT COUNT(*) as count FROM `tabBOM`
    WHERE item_code = %s 
      AND docstatus = 1 
      AND is_active = 1
    """
    
    result = frappe.db.sql(query, (item_code,), as_dict=True)
    return result[0]['count'] > 0

def search_bom_items(search_text):
    """
    Search in BOMs and items
    """
    query = """
    SELECT DISTINCT
        bom.name AS bom_name,
        bom.item_code,
        bom.item_name,
        'BOM' as search_type
    FROM `tabBOM` bom
    WHERE (bom.name LIKE %s OR bom.item_code LIKE %s OR bom.item_name LIKE %s)
      AND bom.docstatus = 1 
      AND bom.is_active = 1
    
    UNION ALL
    
    SELECT DISTINCT
        bom.name AS bom_name,
        bom.item_code,
        bom.item_name,
        'Item' as search_type
    FROM `tabBOM Item` item
    JOIN `tabBOM` bom ON bom.name = item.parent
    LEFT JOIN `tabItem` item_master ON item_master.name = item.item_code
    WHERE (item.item_code LIKE %s OR item.item_name LIKE %s)
      AND bom.docstatus = 1 
      AND bom.is_active = 1
      AND (item_master.item_group = 'محصولات' OR item_master.item_group IS NULL OR item_master.item_group = '')
    
    ORDER BY bom_name
    LIMIT 50
    """
    
    search_param = f"%{search_text}%"
    results = frappe.db.sql(query, (search_param, search_param, search_param, 
                                   search_param, search_param), as_dict=True)
    
    # تبدیل نتایج جستجو به فرمت مناسب
    bom_list = []
    for result in results:
        bom_list.append({
            "name": result['bom_name'],
            "item_code": result['item_code'],
            "item_name": result['item_name'],
            "search_type": result['search_type']
        })
    
    return [{"type": "bom_list", "bom_list": bom_list}]